File: client/src/model/ITransportEvent.ts
import { MessageTypes } from "./MessageTypes";

export interface ITransportEvent {
  sessionId: string;
  payload: Uint8Array<ArrayBufferLike> | null;
  messageType: MessageTypes;
}

export class TransportEvent implements ITransportEvent {
  public sessionId: string;
  public messageType: MessageTypes;
  public payload: Uint8Array<ArrayBufferLike> | null = null;

  constructor(
    sessionId: string, 
    messageType: MessageTypes, 
    payload?: Uint8Array<ArrayBufferLike> | null
  ) {
    this.sessionId = sessionId
    this.messageType = messageType
    if(payload !== null && payload !== undefined) {
      this.payload = payload
    }
  }
}

File: server/Features/ITransportEvent.cs
using MessagePack;

namespace WebTransportExample.Features;

[MessagePackObject(true)]
public class TransportEvent
{
    [Key("sessionId")]
    public required Guid SessionId;

    [Key("messageType")]
    public required MessageTypes MessageType;

    [Key("payload")]
    public byte[]? Payload;
}